tests/integration/data/storage/etcd/store_test.go

```diff
@@ -130,39 +127,52 @@ func TestDelete(t *testing.T) {
 	}
 }
 
-func TestUpdate(t *testing.T) {
-	key := "foo"
-	val := &TestMessage{Id: "100", Name: "bar"}
-	ttl := int64(0)
+func TestConcurrentUpdates(t *testing.T) {
+	ctx := context.Background()
+	key := "entry"
 
-	ctx1, cancel1 := newContext()
-	err := store.Update(ctx1, key, val, ttl)
-	// cancel timeout (release resources) if operation completes before timeout
-	defer cancel1()
-	if err != nil {
-		t.Error(err)
+	// cleanup
+	if err := store.Delete(ctx, key, false, nil); err != nil {
+		if err != storage.NotFound {
+			t.Error(err)
+		}
 	}
 
-	// confirm
-	out := &TestMessage{}
-	ignoreNotFound := false
-	ctx2, cancel2 := newContext()
-	err = store.Get(ctx2, key, out, ignoreNotFound)
-	// cancel timeout (release resources) if operation completes before timeout
-	defer cancel2()
-	if err != nil {
-		t.Error(err)
+	// Create an empty entry
+	value := &TestMessage{}
+	store.Create(ctx, key, value, nil, 0)
+
+	// Perform updates concurrently
+	const iterations = 100
+	var wg sync.WaitGroup
+	wg.Add(iterations)
+	for i := 0; i < iterations; i++ {
+		go func() {
+			defer wg.Done()
+			uf := func(current proto.Message) (proto.Message, error) {
+				msg, ok := current.(*TestMessage)
+				if !ok {
+					return nil, fmt.Errorf("value is not the right type (expected TestMessage): %T", msg)
+				}
+
+				msg.Names = append(msg.Names, fmt.Sprintf("%d", i))
```

Comment:
`i` is not going to be what you expect in your closure. Call your goroutine with a copy of `i` (eg, `go func(i int) {}(i)`).